File: src-tauri/src/commands/tiles.rs
use std::sync::Arc;

use anyhow::Context;
use tauri::State;

use crate::{
    commands::CmdResult,
    database::{
        DbPool, JsonObject,
        entity::{
            folder::FolderId,
            tile::{
                CreateTile, TileIcon, TileIconOptions, TileId, TileLabel, TileModel, TilePosition,
                UpdateKind,
            },
        },
    },
    device::Devices,
    tile::Tiles,
};

/// Get a list of all tiles for a folder
#[tauri::command]
pub async fn tiles_get_tiles(
    db: State<'_, DbPool>,
    folder_id: FolderId,
) -> CmdResult<Vec<TileModel>> {
    let tiles = TileModel::get_by_folder(db.inner(), folder_id).await?;
    Ok(tiles)
}

/// Get a specific tile
#[tauri::command]
pub async fn tiles_get_tile(
    db: State<'_, DbPool>,
    tile_id: TileId,
) -> CmdResult<Option<TileModel>> {
    let tiles = TileModel::get_by_id(db.inner(), tile_id).await?;
    Ok(tiles)
}

/// Create a new tile
#[tauri::command]
pub async fn tiles_create_tile(
    db: State<'_, DbPool>,
    devices: State<'_, Arc<Devices>>,
    create: CreateTile,
) -> CmdResult<TileModel> {
    let db = db.inner();
    let tile = TileModel::create(db, create).await?;
    devices.background_update_folder(tile.folder_id);
    Ok(tile)
}

/// Update a specific tile properties
#[tauri::command]
pub async fn tiles_update_tile_properties(
    tiles: State<'_, Arc<Tiles>>,
    tile_id: TileId,
    properties: JsonObject,
    partial: bool,
) -> CmdResult<TileModel> {
    let tile = tiles
        .update_tile_properties(tile_id, None, properties, partial)
        .await?;
    Ok(tile)
}

/// Update a specific tile label
#[tauri::command]
pub async fn tiles_update_tile_label(
    tiles: State<'_, Arc<Tiles>>,
    tile_id: TileId,
    label: TileLabel,
    kind: UpdateKind,
) -> CmdResult<TileModel> {
    let tile = tiles.update_tile_label(tile_id, None, label, kind).await?;

    Ok(tile)
}
/// Update a specific tile label
#[tauri::command]
pub async fn tiles_update_tile_position(
    tiles: State<'_, Arc<Tiles>>,
    tile_id: TileId,
    position: TilePosition,
) -> CmdResult<TileModel> {
    let tile = tiles.update_tile_position(tile_id, None, position).await?;

    Ok(tile)
}

/// Update a specific tile icon
#[tauri::command]
pub async fn tiles_update_tile_icon(
    tiles: State<'_, Arc<Tiles>>,
    tile_id: TileId,
    icon: TileIcon,
    kind: UpdateKind,
) -> CmdResult<TileModel> {
    let tile = tiles.update_tile_icon(tile_id, None, icon, kind).await?;
    Ok(tile)
}

/// Update a specific tile icon options
#[tauri::command]
pub async fn tiles_update_tile_icon_options(
    tiles: State<'_, Arc<Tiles>>,
    tile_id: TileId,
    icon_options: TileIconOptions,
) -> CmdResult<TileModel> {
    let tile = tiles
        .update_tile_icon_options(tile_id, None, icon_options)
        .await?;
    Ok(tile)
}

/// Delete a specific tile
#[tauri::command]
pub async fn tiles_delete_tile(
    db: State<'_, DbPool>,
    devices: State<'_, Arc<Devices>>,
    tile_id: TileId,
) -> CmdResult<()> {
    let db = db.inner();
    let tile = TileModel::get_by_id(db, tile_id)
        .await?
        .context("tile not found")?;

    TileModel::delete(db, tile_id).await?;
    devices.background_update_folder(tile.folder_id);
    Ok(())
}

File: src/lib/api/tiles.ts
import { invoke } from "@tauri-apps/api/core";
import { createQuery, createMutation } from "@tanstack/svelte-query";

import type { FolderId } from "./types/folders";
import type {
  TileId,
  TileIcon,
  TileModel,
  TileLabel,
  CreateTile,
  UpdateKind,
} from "./types/tiles";

import { queryClient } from "./client";
import { runeStore } from "./utils/svelte.svelte";

export const tilesKeys = {
  root: ["tiles"],
  list: (folderId: FolderId | null) => ["tiles", folderId, "list"],
  specific: (folderId: FolderId | null, tileId: TileId | null) => [
    "tiles",
    folderId,
    "specific",
    tileId,
  ],
};

// [REQUESTS] ------------------------------------------------------

function getTiles(folderId: FolderId) {
  return invoke<TileModel[]>("tiles_get_tiles", { folderId });
}

export function getTile(tileId: TileId) {
  return invoke<TileModel | null>("tiles_get_tile", { tileId });
}

function createTile(create: CreateTile) {
  return invoke<TileModel>("tiles_create_tile", {
    create,
  });
}

function updateTileProperties(
  tileId: TileId,
  properties: object,
  partial: boolean,
) {
  return invoke<TileModel>("tiles_update_tile_properties", {
    tileId,
    properties,
    partial,
  });
}

function updateTileLabel(tileId: TileId, label: TileLabel, kind: UpdateKind) {
  return invoke<TileModel>("tiles_update_tile_label", {
    tileId,
    label,
    kind,
  });
}

function updateTileIcon(tileId: TileId, icon: TileIcon, kind: UpdateKind) {
  return invoke<TileModel>("tiles_update_tile_icon", {
    tileId,
    icon,
    kind,
  });
}

function deleteTile(tileId: TileId) {
  return invoke("tiles_delete_tile", { tileId });
}

// [QUERIES] ------------------------------------------------------

export function createTilesQuery(folderId: () => FolderId) {
  return createQuery(
    runeStore(() => {
      const id = folderId();
      return {
        queryKey: tilesKeys.list(id),
        queryFn: () => getTiles(id!),
      };
    }),
  );
}

export function createTileQuery(
  folderId: () => FolderId,
  tileId: () => TileId,
) {
  return createQuery(
    runeStore(() => {
      const fid = folderId();
      const tid = tileId();
      return {
        queryKey: tilesKeys.specific(fid, tid),
        queryFn: () => getTile(tid!),
      };
    }),
  );
}

// [MUTATIONS] ------------------------------------------------------

export function createCreateTileMutation() {
  return createMutation({
    mutationFn: ({ create }: { create: CreateTile }) => createTile(create),
    onSuccess: (tile) => {
      invalidateTilesList(tile.folder_id);
      queryClient.setQueryData(
        tilesKeys.specific(tile.folder_id, tile.id),
        tile,
      );
    },
  });
}

export function createUpdateTilePropertiesMutation() {
  return createMutation({
    scope: { id: "tile" },
    mutationFn: ({
      tileId,
      properties,
      partial,
    }: {
      tileId: TileId;
      properties: object;
      partial: boolean;
    }) => updateTileProperties(tileId, properties, partial),
    onSuccess: (tile) => {
      invalidateTilesList(tile.folder_id);
      queryClient.setQueryData(
        tilesKeys.specific(tile.folder_id, tile.id),
        tile,
      );
    },
  });
}

export function createUpdateTileLabelMutation() {
  return createMutation({
    scope: { id: "tile" },
    mutationFn: ({
      tileId,
      label,
      kind,
    }: {
      tileId: TileId;
      label: TileLabel;
      kind: UpdateKind;
    }) => updateTileLabel(tileId, label, kind),
    onSuccess: (tile) => {
      invalidateTilesList(tile.folder_id);
      queryClient.setQueryData(
        tilesKeys.specific(tile.folder_id, tile.id),
        tile,
      );
    },
  });
}

export function createUpdateTileIconMutation() {
  return createMutation({
    scope: { id: "tile" },
    mutationFn: ({
      tileId,
      icon,
      kind,
    }: {
      tileId: TileId;
      icon: TileIcon;
      kind: UpdateKind;
    }) => updateTileIcon(tileId, icon, kind),
    onSuccess: (tile) => {
      invalidateTilesList(tile.folder_id);
      queryClient.setQueryData(
        tilesKeys.specific(tile.folder_id, tile.id),
        tile,
      );
    },
  });
}

export function createDeleteTileMutation(folderId: FolderId) {
  return createMutation({
    mutationFn: ({ tileId }: { tileId: TileId }) => deleteTile(tileId),
    onSuccess: (_, { tileId }) => {
      invalidateTilesList(folderId);
      queryClient.removeQueries({
        queryKey: tilesKeys.specific(folderId, tileId),
      });
    },
  });
}

// [MUTATORS] ------------------------------------------------------

function invalidateTilesList(folderId: FolderId) {
  queryClient.invalidateQueries({
    queryKey: tilesKeys.list(folderId),
    exact: false,
  });
}
